src/mscorlib/shared/System/Memory.cs

```diff
@@ -0,0 +1,313 @@
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> is not in the range (&lt;0 or &gt;=Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory(T[] array, int start)
+        {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            if (default(T) == null && array.GetType() != typeof(T[]))
+                ThrowHelper.ThrowArrayTypeMismatchException();
+
+            int arrayLength = array.Length;
+            if ((uint)start > (uint)arrayLength)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            _arrayOrOwnedMemory = array;
+            _index = start;
+            _length = arrayLength - start;
+        }
+
+        /// <summary>
+        /// Creates a new memory over the portion of the target array beginning
+        /// at 'start' index and ending at 'end' index (exclusive).
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <param name="start">The index at which to begin the memory.</param>
+        /// <param name="length">The number of items in the memory.</param>
+        /// <exception cref="System.ArgumentNullException">Thrown when <paramref name="array"/> is a null
+        /// reference (Nothing in Visual Basic).</exception>
+        /// <exception cref="System.ArrayTypeMismatchException">Thrown when <paramref name="array"/> is covariant and array's type is not exactly T[].</exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> or end index is not in the range (&lt;0 or &gt;=Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory(T[] array, int start, int length)
+        {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            if (default(T) == null && array.GetType() != typeof(T[]))
+                ThrowHelper.ThrowArrayTypeMismatchException();
+            if ((uint)start > (uint)array.Length || (uint)length > (uint)(array.Length - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            _arrayOrOwnedMemory = array;
+            _index = start;
+            _length = length;
+        }
+        
+        // Constructor for internal use only.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Memory(OwnedMemory<T> owner, int index, int length)
+        {
+            _arrayOrOwnedMemory = owner;
+            _index = index | (1 << 31); // Before using _index, check if _index < 0, then 'and' it with bitMask
+            _length = length;
```

Comment:
Don't we still need to validate arguments?  OwnedMemory's Length is abstract, so a derived OwnedMemory could, for example, return a negative value out of Length, which is then passed into this Memory ctor and stored here.

Reply:
This is an internal constructor. I don't follow why the argument validation is necessary (which will add overhead to Slice calls).

I could add a null check for owner (and index/length >= 0), but what type of validation are you suggesting? We aren't accessing OwnedMemory.Length property here.

Reply:
> This is an internal constructor.

It's called by OwnedMemory here:
https://github.com/dotnet/coreclr/pull/13583/files#diff-0184445324edc32335f6249267a0bf70R29
and there's no validation done on the Length, which provides an arbitrary result based on whatever the derived type chose to return:
https://github.com/dotnet/coreclr/pull/13583/files#diff-0184445324edc32335f6249267a0bf70R14
which means I can write an OwnedMemory-derived type that returns -42 for Length, and then get a Memory that has a negative length.